### DISCUSSION TITLE
How to use the router level dependency in the routes?

### DISCUSSION THREAD

**Discussion by abbazs** [fastapi/fastapi#8810] [Questions] [answered]
### Example Code

```python
def get_db_connection():
    conn = pymssql.connect(
        server=settings.server,
        user=settings.user,
        password=settings.password,
        database=settings.database,
        port=settings.port,
        as_dict=True,
    )
    try:
        yield conn
    finally:
        conn.close()


router = APIRouter(
    prefix="/project",
    tags=["project"],
    dependencies=[Depends(get_db_connection)],
    responses={418: {"description": "Project status API."}},
)


@router.get("/count/{status}")
def projects_count(
    status: str, conn: pymssql.Connection = Depends(get_db_connection)
) -> dict:
    try:
        if status.upper() == "ALL":
            status = None
        return aq.get_project_count(conn, status=status)
    except:
        raise HTTPException(404, "Unable to get project count")
```

### Description

There are several endpoints in the router. And all of them have the parameter conn. Can I avoid declaring this parameter at each router by declaring it as the dependency at APIRouter? And how to access the dependency declared at APIRouter in a route?

### Operating System

Linux

### FastAPI Version

0.66.0

### Python Version

3.9.6

### Additional Context

When I did google search for the same, the following is what I get:
https://github.com/tiangolo/fastapi/issues/424
https://stackoverflow.com/questions/63961339/how-to-access-route-dependencies-in-custom-apiroute-class
https://stackoverflow.com/questions/64506036/fastapi-how-to-access-attributes-of-class-based-dependency-inside-route-when-dep

Is there a simpler solution?

**Comment by jiteshy** [comment]
The way you have specified dependencies in APIRouter as below is correct (Reference: [Dependency in APIRouter](https://fastapi.tiangolo.com/tutorial/bigger-applications/#add-some-custom-tags-responses-and-dependencies) ) :
`dependencies=[Depends(get_db_connection)]`

This can be taken even one more level up to declaring FastAPI app if it's needed for all the Routers (Reference: [Dependency in FastAPI](https://fastapi.tiangolo.com/tutorial/dependencies/global-dependencies/) )
`app = FastAPI(dependencies=[Depends(get_db_connection)])`

With that, I think accessing the dependency as below should work considering you are specifying the dependency in the router (It worked for my sample code).

`router.dependencies[0].dependency()`

however I am not sure if there is a more elegant way to access the Router/App level dependencies in endpoints.

**Comment by ben519** [comment]
I also have this question. Additionally..

Is there any benefit to including `dependencies=[Depends(get_db_connection)]` within the declaration of `APIRouter()`, considering we need to include it as a parameter in every path operation function? Seems unnecessarily redundant.

**Reply by juliocefe** [reply]
Same here 

Version: fastapi[all]==0.95.1

Problem:
I am also stuck with this issue. **I have declared Depends(get_db) in my router dependencies list, but I need to include the same Depends declaration in every route. Otherwise, FastAPI throws an error**:
fastapi.exceptions.FastAPIError: Invalid args for response field! Hint: check that <class 'sqlalchemy.orm.session.Session'> is a valid Pydantic field type. more...

```
router = APIRouter(
    prefix="/users",
    dependencies=[Depends(get_db)],
)

# This declaration works
@router.get("/")
async def get_users(db: Session = Depends(get_db)) -> UserScheema:
    return crud.read_users(db)

# But, when I use this approach, FastAPI raises the aforementioned error.
@router.get("/")
async def get_users(db: Session) -> UserScheema:
    return crud.read_users(db)
```

Thank you for this amazing framework, I am loving the documentation ❤️.

**Reply by juliocefe** [reply]
This solved my problem!
https://fastapi.tiangolo.com/tutorial/dependencies/#share-annotated-dependencies
Great!

**Reply by will-e-yams** [reply]
Adding the `Depends` to the endpoint as a "fix" just proves the point of this bug.

**Comment by marcodelmoral** [comment]
Any ideas how to reduce redundancy in the declaration of router dependencies? I want to access my settings dependency, usually declared as a global dependency, inside my router endpoints.

**Comment by wesselvl1** [comment]
```py
router = APIRouter(
    prefix="/il_oedometer/casagrande",
    tags=["/il_oedometer/casagrande"],
    dependencies=[Depends(deps.get_casagrande_service)],
)

logger = logging.getLogger(__name__)
casagrandeService = deps.get_casagrande_service()
```

I actually did it like this so the casagrandeService is callable in all functions. But came here to figure out if this was a good way to tackle this or not. (Probably not)

**Reply by will-e-yams** [reply]
I don't see how this solves anything. Even without the router dependency `casagrandeService` would be available to any code below this.

**Comment by will-e-yams** [comment]
I'm guessing no one figured this out and we have to put dependencies on each. and. every. endpoint.

Maybe I can convince someone to "Depend" all the routes and I can make it middleware.

**Reply by arielnmz** [reply, marked as answer]
Re:

> I don't see how this solves anything.

I think the solution looks like this:

```
router = APIRouter(prefix="/items", tags=["items"])

DBSession = Annotated[Session, Depends(get_db)]

@router.get("/")
async def customers_list(db: DBSession) -> list[Item]:
    return db.do_something()
```
Two things:
1. Notice how I skipped `dependencies=[Depends(get_db)]` from `APIRouter` and instead declared annotated `Session` with the dep.
2. Notice how you don't have to do `db: Session = Depends(get_db)`  on every route handler.

Imo seems counter intuitive, but I think the magic was that `Session` being `Annotated` makes FastAPI inject it to the route handler.

**Reply by will-e-yams** [reply]
@arielnmz sweet solution!

![clever-girl-memegenerator-net-34396869.png](https://github.com/user-attachments/assets/55c850dc-cbf9-4923-9af8-ac7b55221770)

**Comment by DJoepie** [comment]
what is the point of `dependencies=[Depends(some.dep)],` if you can't actually use it on the endpoints?
For example, I need to pass my db as a dep on almost all routes in 1 router file but I can't seem to figure out how I can just transform
`def read_user(user_id: int, db: Session = Depends(get_db)):`
into
`def read_user(user_id: int, db):`

**Reply by n8sty** [reply]
The point of router level dependencies is that there are types of dependencies that need to be run as part of a route but do not themselves need to be accessed within a route. For example, take an authentication dependency that checks for the presence of the correct credentials provided in either a header or cookie or through some other method as part of a request and raises a 401 exception when authorization fails. Once the authorization has been checked it's possible, depending on the design of your application, that you no longer need anything provided by that dependency, but it is important that it runs to protect a given route.

**Reply by DJoepie** [reply]
ohhh so if I understand you correctly I can protect all routes within the admin router with `is_admin` since I'm simply checking the authorization.

But if I want to do something with i.e. a database, I would still have to do that on a per route basis since I'm actively using the db "instance".

Do I understand what you're saying correctly like this?

**Reply by will-e-yams** [reply]
Afaik, you are correct. But damn, I wish that we could inject dependencies at the router level.

Say I have admin router and I want to perform additional scrutiny:

Currently:
1. Add dependency to the router which keeps them out
2. Add a dependency to the endpoint that gives me access to the exact same data (to know who they are)

It's not the end of the world, but it is more boilerplate (and probably compute)

**Reply by n8sty** [reply]
@DJoepie Yep, you're understanding correctly.

**Reply by nfnfgo** [reply]
Router-level dependencies is limited in their functionality if we have no way to access the return value of the deps especially when we need to react based on the result of those dependencies 😭

**Reply by fkapsahili** [reply]
Would a PR be reviewed for this? 🙂

**Comment by mammothrider** [comment]
edit: fix format

maybe adding a dict_dependencies and pass that as a default param to endpoints.

for example:

```
router = APIRouter(
    prefix="/project",
    tags=["project"],
    dict_dependencies=dict(
        conn = Depends(get_db_connection)
    ),
    responses={418: {"description": "Project status API."}},
)


@router.get("/count/{status}")
def projects_count(
    status: str, conn: pymssql.Connection # <- should get dep injection
) -> dict:
    try:
        if status.upper() == "ALL":
            status = None
        return aq.get_project_count(conn, status=status)
    except:
        raise HTTPException(404, "Unable to get project count")
```

**Reply by will-e-yams** [reply]
plz use triple backticks for code blocks, per markdown standards. single backticks are for code contained in a sentence.

that said, what is `aq` in this sample? (note the use of single backtick here)

ah, i see that you carried that over from the OP's question. i'm going to assume is a module for DB access that supports `get_project_count`.

A/ does this actually work?
B/ does it integrate with the OpenAPI docs?

this is what the docs say now, which while not a router dependency injection, is, at least, a bit more reusable:

https://fastapi.tiangolo.com/tutorial/dependencies/#create-a-dependency-or-dependable

```
from typing import Annotated

from fastapi import Depends, FastAPI

app = FastAPI()


async def common_parameters(q: str | None = None, skip: int = 0, limit: int = 100):
    return {"q": q, "skip": skip, "limit": limit}


@app.get("/items/")
async def read_items(commons: Annotated[dict, Depends(common_parameters)]):
    return commons


@app.get("/users/")
async def read_users(commons: Annotated[dict, Depends(common_parameters)]):
    return commons
```

being repurposed into injecting a local db connection as a parameter rather than query string params.

THIS IS AN UNTESTED APPROACH.

**Reply by mammothrider** [reply]
> plz use triple backticks for code blocks, per markdown standards. single backticks are for code contained in a sentence.
> 
> that said, what is `aq` in this sample? (note the use of single backtick here)
> 
> ah, i see that you carried that over from the OP's question. i'm going to assume is a module for DB access that supports `get_project_count`.
> 
> A/ does this actually work? B/ does it integrate with the OpenAPI docs?
> 
> this is what the docs say now, which while not a router dependency injection, is, at least, a bit more reusable:
> 
> https://fastapi.tiangolo.com/tutorial/dependencies/#create-a-dependency-or-dependable
> 
> ```
> from typing import Annotated
> 
> from fastapi import Depends, FastAPI
> 
> app = FastAPI()
> 
> 
> async def common_parameters(q: str | None = None, skip: int = 0, limit: int = 100):
>     return {"q": q, "skip": skip, "limit": limit}
> 
> 
> @app.get("/items/")
> async def read_items(commons: Annotated[dict, Depends(common_parameters)]):
>     return commons
> 
> 
> @app.get("/users/")
> async def read_users(commons: Annotated[dict, Depends(common_parameters)]):
>     return commons
> ```
> 
> being repurposed into injecting a local db connection as a parameter rather than query string params.
> 
> THIS IS AN UNTESTED APPROACH.

sry, format fixed.

for me, what the router dependencies can only do, is user verification. and for many endpoints under this router, there are some common dependecies, eg,  get sql session. 
Currently, I need to add a sql depends in every endpoint, and I think this could be add to router level.

PS:
I'm also consider directly getting the data in depends and doing verify, instead of getting sql session. And I'm still not sure if this is a good practice.
